### ISSUE TITLE
Attribute routing has stopped working

### ISSUE THREAD

**Issue by jcrittenden** [umbraco/Umbraco-CMS#16816]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

13.4.1

### Bug summary

There is something in versions >=13.3.1 that broke attribute routing. For instance, I have a UmbracoPageController and the controller applies to all pages versus just the "filtered" routes.

Example:
[Host("dev.mysite.com", "qa.mysite.com")]
public class MySiteContoller : UmbracoPageController, IVirtualPageController {.....

My temporary solution was to not upgrade past 13.3.0, but the issue persists in 13.4.1 (the newest LTS minor).

### Steps to reproduce

Apply a route attribute to a page controller, and it will apply to all paths/hosts instead of the filtered ones.

---
_This item has been added to our backlog AB#50831_

**Comment by Migaroez** [comment]
Hey @jcrittenden, I have tried to reproduce your issue with the limited reproduction steps you supplied but have been unsuccessful.
In the video below you can see that I am able to add the Host attribute to a slightly modified version of the IVirtualPageController example from our documentation. Below that you can find all the code I added to the solution.

Findings:
Not supplying the Host attribute,  results in the route /products/id hitting the controller
Adding the host with my localhost domain, results in the route /products/id hitting the controller
Removing the localhost domain from the host attribute but leaving it present with other domains, results in the route not being hit

This seems to be in line with what you expect


https://github.com/user-attachments/assets/39408032-5815-4e6d-afd6-16d85d582545

ProductController and models
```csharp
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Controllers;
using Microsoft.AspNetCore.Mvc.Filters;
using Microsoft.AspNetCore.Mvc.ViewEngines;
using Umbraco.Cms.Core.Models;
using Umbraco.Cms.Core.Models.PublishedContent;
using Umbraco.Cms.Core.Web;
using Umbraco.Cms.Web.Common.Controllers;

namespace ShouldNeverBeDeployed;

[Host("dev.mysite.com", "qa.mysite.com")]
public class ProductController : UmbracoPageController, IVirtualPageController
{
    private readonly IUmbracoContextAccessor _umbracoContextAccessor;

    public ProductController(
        ILogger<UmbracoPageController> logger,
        ICompositeViewEngine compositeViewEngine,
        IUmbracoContextAccessor umbracoContextAccessor
    )
        : base(logger, compositeViewEngine)
    {
        _umbracoContextAccessor = umbracoContextAccessor;
    }

    public IPublishedContent? FindContent(ActionExecutingContext actionExecutingContext)
    {
        var umbracoContext = _umbracoContextAccessor.GetRequiredUmbracoContext();

        var productRoot = umbracoContext.Content?.GetAtRoot().First();
        if (productRoot == null)
        {
            return null;
        }

        if (actionExecutingContext.ActionDescriptor is not ControllerActionDescriptor controllerActionDescriptor)
        {
            return productRoot;
        }
        // Check which action is executing
        return controllerActionDescriptor.ActionName switch
        {
            nameof(Index) => productRoot,
            nameof(Product) => productRoot.Children.First(),
            _ => productRoot
        };
    }

    [HttpGet]
    [Route("products")]
    public IActionResult Index()
    {
        return View("ProductArea",CurrentPage);
    }

    [HttpGet]
    [Route("products/{id}")]
    public IActionResult Product(string id)
    {
        var productItemModel = new ProductModel(CurrentPage)
        {
            Sku = id,
            NotUmbracoData = "This is some extra data",
        };

        return View("ProductDisplay", productItemModel);
    }

    public class ProductModel : ContentModel
    {
        public ProductModel(IPublishedContent content) : base(content)
        {
        }

        public string Sku { get; set; }
        public string NotUmbracoData { get; set; }
    }
}
```

ProductArea.cshtml
```cshtml
@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage<ContentModels.ProductArea>
@using ContentModels = Umbraco.Cms.Web.Common.PublishedModels;
@{
	Layout = null;
}

@Model.Rich
```

ProductDisplay.cshtml
```cshtml
@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage<ShouldNeverBeDeployed.ProductController.ProductModel>
@{
	Layout = null;
	var typedUmbracoModel = (ProductDisplay)Model.Content;
}
<h1>@Model.Sku</h1>
@typedUmbracoModel.Disclaimer
<p>@Model.NotUmbracoData</p>
```

**Comment by jcrittenden** [comment]
@Migaroez You need to have a multi-site set up. So for instance, site1.localhost and site2.localhost.

So the given trees are:

site1
    productpage
site2
    productpage
    
Make the controller ONLY apply to the product pages of site1.localhost in the host filter. Now attempt to go to "site2.localhost/productpage" You should get to the page with default routing. However, the controller will be applied to the other domain as well.

**Comment by jcrittenden** [comment]
@Migaroez Please take another look when you can please.

**Comment by jcrittenden** [comment]
[MyRoutingController.txt](https://github.com/user-attachments/files/17546977/MyRoutingController.txt)
[MyRoutingComposer.txt](https://github.com/user-attachments/files/17546979/MyRoutingComposer.txt)

@Migaroez Thank you for looking again. Could it be my Composer? I've attached my composer and controller. But I would be happy to see what you have as well, that is working as expected.

**Comment by Migaroez** [comment]
Yeah, I don't think you need a composer. You either use routing attributes or you define the routing with app.UseEndpoints.

The app.UserEndpoints does not know about the host you configured on the controller as that is read out by another process to combine alle the different attributes and then add all unique urls to the routing table. (shortened explanation and there might be some nuances I am missing)

The docs hint at this but might be a bit clearer.
https://docs.umbraco.com/umbraco-cms/reference/routing/custom-routes#attribute-routing-with-ivirtualpagecontroller

> ...If we want to convert our above example into using attribute routing...

Convert implies not using them at the same time, but I can understand the confusion.

The .net docs about mvc routing (which Umbraco uses) are a lot clearer about it
https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-8.0

> Actions are either [conventionally-routed](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-8.0#cr6) or [attribute-routed](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-8.0#ar6). Placing a route on the controller or [action](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-8.0#action) makes it attribute-routed. See [Mixed routing](https://learn.microsoft.com/en-us/aspnet/core/mvc/controllers/routing?view=aspnetcore-8.0#routing-mixed-ref-label) for more information.

So try to remove the composer as the attributes seem to hold the correct information and see how it behaves.

**Comment by nielslyngsoe** [comment]
Hi @jcrittenden did the conversation with @Migaroez help you solve the problem or do you still see this issue begin present?
Please let me know, otherwise I will close the issue.

Thanks in advance

**Comment by Migaroez** [comment]
Hey @jcrittenden here is the zip file of how I made it work 
[issue16816.zip](https://github.com/user-attachments/files/18344569/issue16816.zip)

**Comment by jcrittenden** [comment]
> Hey [@jcrittenden](https://github.com/jcrittenden) here is the zip file of how I made it work [issue16816.zip](https://github.com/user-attachments/files/18344569/issue16816.zip)

Okay, @Migaroez your zip example did in-fact work as expected. However, it was set up differently, **without the route being at the first level**. ie. [Route("products/{id}")] instead of [Route("{id}")]

I've simplified and edited your example to illustrate the issue. Attached here:

[issue16816Revised.zip](https://github.com/user-attachments/files/19324863/issue16816Revised.zip)

Username: test@test.test
Password: password123

Note that the controller should ONLY apply to "site1." However, try using these urls:
https://site1.localhost:44323/      (Loads site1 **_as expected_**)
https://site2.localhost:44323/      (Loads site2 **_as expected_**)
https://site1.localhost:44323/1060/      (Uses controller, finds product, and displays page **_as expected_**)
https://site1.localhost:44323/product-display/      (Actual page, but uses controller and 404s **as expected**)

https://site2.localhost:44323/product-display-site-two/      **This is an actual page in the tree on site2. So, route controller should not affect it. Yet... we get a 404...   _NOT as expected_**

![Image](https://github.com/user-attachments/assets/f310fa9d-0c23-42dd-bc82-9078dac91a84)

This absolutely/100% works as-is in Umbraco versions <=13.3.0. And we haven't been able to upgrade because of it. Urgency has increased with the vulnerabilities. @nielslyngsoe

**Comment by jcrittenden** [comment]
@AndyButland Thank you so very much for the help here!! Please let me know what release this will be on, when that is known.

**Comment by AndyButland** [comment]
Will do @jcrittenden - we'll update here with the release once it's reviewed, tested and merged.

If you were happy verifying from source code and wanted to try it out yourself in your setup, please do as that would be valuable.  Just to be sure we haven't missed something in replicating the issue and so we can be confident it'll resolve your issue when released.

**Comment by AndyButland** [comment]
Fixed via https://github.com/umbraco/Umbraco-CMS/pull/18820 and will be included in 13.8.
